project_generator/project.py

```diff
@@ -124,139 +125,191 @@ def _process_source_files(self, files, group_name):
+                project._set_output_dir_path(export_tool, workspace_path)
 
-        self.name = name
+                project._set_output_dir()
+                if copy:
+                    project.copy_sources_to_generated_destination
+                project.project['singular'] = False
+                files = exporter(project.project, self.pgen_workspace.settings).export_project()
+                # we gather all generated files, needed for workspace files
+                workspace_dic['projects'].append(files)
+                generated_files['projects'].append(files)
 
-        self.include_paths = []
-        self.source_paths = []
-        self.source_groups = {}
-        self.macros = []
-        self.project_dir = {
-            'name': '',
-            'path': '',
-        }
-        self.build_dir = 'build'
-        self.output_dir = {
-            'path': '',
-            'rel_path': '',
-            'rel_count': '',
-        }
-        self.mcu = ''
-        self.core = ''
-        self.target = ''
-        self.tools_supported = []
+            # all projects are genereated, now generate workspace files
+            generated_files['workspaces'] = exporter(workspace_dic, self.pgen_workspace.settings).export_workspace()
+
+            self.generated_files[export_tool] = generated_files
+
+class Project:
+
+    """represents a project, which can be formed of many yaml files"""
 
+    def __init__(self, name, project_dicts, pgen_workspace):
+        """initialise a project with a yaml file"""
+        self.pgen_workspace = pgen_workspace
+        self.tool_specific = defaultdict(ToolSpecificSettings)
+        self.name = name
         self.output_types = {
             'executable': 'exe',
             'exe': 'exe',
             'library': 'lib',
             'lib': 'lib',
         }
-        self.output_type = self.output_types['executable']
-
-        self.debugger = 'cmsis-dap'
+        self.tools = ToolsSupported()
+        self.source_groups = {}
+        self.project = {}
+        self._fill_project_defaults()
+        # process all projects dictionaries
+        for project in project_dicts:
+            self._set_project_attributes(project)
+        self.generated_files = {}
+
+    def _fill_project_defaults(self):
+
+        self.project = {
+            'name': self.name,          # project name
+            'core': '',                 # core
+            'linker_file': None,        # linker command file
+            'build_dir' : 'build',      # Build output path
+            'debugger' : 'cmsis-dap',   # Debugger
+            'includes': [],             # include paths
+            'copy_sources': False,      # [internal] Copy sources to destination flag
+            'include_files': [],        # [internal] files to be included
+            'source_paths': [],         # [internal] source paths
+            'source_files_c': [],       # [internal] c source files
```

Comment:
Because you can't set them in the yaml file. We should distinguish between what can be imported from a user and what pgen generates as internal data.